src/db.rs

```diff
@@ -1115,6 +1115,28 @@ impl Db {
     pub fn metrics(&self) -> Arc<DbStats> {
         self.inner.db_stats.clone()
     }
+
+    /// Only used for test cases in slateDB internal, please don't use it as a normal operation
+    #[allow(unused)]
+    async fn force_flush(&self) -> Result<(), SlateDBError> {
```

Comment:
why can't we just use `flush` to do this? I guess one obvious reason is that `flush` doesn't freeze the memtable when the wal is disabled, but that just seems like a problem with the implementation of flush that we should fix.

Reply:
I checked the implementation of flush, `flush` does not freeze the current WAL and memtable. Since `flush` is a public API, I'm not sure whether to modify it.

Reply:
But I’m fine with both options, whether it’s modifying the original `flush` or introducing a new auxiliary function only for testing.

Reply:
So in the current impl `flush` does actually freeze the current WAL table - it's done by the flush thread when processing a `FlushImmutableWals` (not a great name) message. It does not freeze the memtable when wal is not enabled. This seems like a bug - a user that does a `put_with_options` with `await_durable: false` followed by a `flush` should expect the the data from the `put_with_options` to be durably written after `flush` returns. This is not the case today because the data would still be in the live memtable. So I think the right approach would be to have `flush` call `freeze_memtable` before telling the memtable flush thread to flush the memtables.

Reply:
Sorry for the late reply, I checked the current implementation. Currently `DB::flush` performs the following logic:
1. If WAL is enabled, we only flush immutable wals to the active memtable.
2. If WAL is not enabled, we flush the immutable memtables and persist them.

I agree that we can try to fix the logic of DB::flush instead of introducing a new function. But I think we need to make a key point in the comment:
When the user calls `DB::flush` at a certain moment, flush will freeze the current active memtable and WAL, and then flush them in the background. This means that foreground writes will not be blocked later, because there is a new active memtable and WAL to use. I think we need to note this in the documentation.

Reply:
In addition, I think it is worth mentioning that `DBInner::flush`. I think its naming is not reasonable. According to its implementation, it will only freeze the current active WAL and then flush it to memtable. I think it is more appropriate to rename it to `DBInner::flush_wals`.
[SlateDB DBInner::flush](https://github.com/slatedb/slatedb/blob/main/src/flush.rs#L26)

Reply:
>  think it is more appropriate to rename it to DBInner::flush_wals.

I agree with your point, but `DbInner.flush_wals` already exists:

https://github.com/slatedb/slatedb/blob/fcb703b196215093cca29a3963b7ab7d9e58f5f3/src/db.rs#L413

What about something like "flush_current_wal"?

Reply:
I think some of this ugliness is just 'cuz we have all the methods in `DbInner`. `mem_table_flush.rs` is a bit cleaner since it pulled these methods into a separate struct. Perhaps that's what we should do.

Reply:
> So I think the right approach would be to have flush call freeze_memtable before telling the memtable flush thread to flush the memtables.

WDYT about freezing the memtable in the flush thread instead? This seems a bit safer, and is how flush.rs works for WALs.

Reply:
> WDYT about freezing the memtable in the flush thread instead?

I think it's a good idea, we can introduce a function like `flush_all_memtables`
```
async fn flush_all_memtables(args) -> Result<(), SlateDBError> {
     DBInner::state.write().freeze_memtable(wal_id)?;
     let result = flusher.flush_imm_memtables_to_l0().await;
}
```
BTW, how about using `DBInner::flush_all_wals` instead of `DBInner::flush`?

Reply:
> Perhaps that's what we should do.

I think this is worth opening a new issue.

Reply:
If we can reach a consensus on fixing `DB::flush`, I can start working on it over the weekend.

Reply:
How about limiting this PR to fixing the behavior of `DB::flush`? We can put renaming and other things in the new issue, which involves refactoring some of the current flush logic.

Reply:
Sounds good!

Reply:
> WDYT about freezing the memtable in the flush thread instead? This seems a bit safer, and is how flush.rs works for WALs.

Yep, makes sense to me.

> I think it's a good idea, we can introduce a function like flush_all_memtables
async fn flush_all_memtables(args) -> Result<(), SlateDBError> {
     DBInner::state.write().freeze_memtable(wal_id)?;
     let result = flusher.flush_imm_memtables_to_l0().await;
}
BTW, how about using DBInner::flush_all_wals instead of DBInner::flush?

I think for memtable flushing it still makes sense for the freezing to be done from the client or wal-flush task and the flushing to be handled by the memtable-flush task. The memtable is frozen based on a size constraint rather than at an interval, so the task that's writing to memtable is the most reasonable place to decide when to freeze, and then it can notify the memtable-flush task that it should flush.

For wal, since the flush task is both freezing based on an interval an flushing, I think it makes sense to have flush just pass it a message to tell it to both freeze and flush.

Reply:
> WDYT about freezing the memtable in the flush thread instead? This seems a bit safer, and is how flush.rs works for WALs.
Yep, makes sense to me.

Sorry, I misunderstood - I thought we were talking about the WAL tables here for some reason. For the memtable it seems fine to freeze from `flush` directly. `freeze_memtable` is expected to be called from both writing tasks and the wal-flush task, so I don't see a synchronization problem. And moving freezing to the memtable flush task will be awkward. The memtable flush task doesn't currently freeze the memtable so we'd have to either:
(1) change the regular path to have the memtable flush task be the one that freezes the memtable. This will likely result in us creating slightly larger memtables since currently freezing is synchronized with the writes to the memtable.
(2) pass some flag to the memtable flush task indicating whether it should freeze or not before flushing. Not the worst thing in the world but seems simpler to just freeze before sending the flush msg.

Reply:
I think the flush of memtable should be consistent with the behavior of WAL. WAL will freeze the current active WAL after receiving the channel message. [DBInner::flush_memtables](https://github.com/slatedb/slatedb/blob/main/src/db.rs#L421)

https://github.com/slatedb/slatedb/blob/main/src/flush.rs#L26
https://github.com/slatedb/slatedb/blob/main/src/flush.rs#L133

Reply:
> I think the flush of memtable should be consistent with the behavior of WAL. WAL will freeze the current active WAL after receiving the channel message. [DBInner::flush_memtables](https://github.com/slatedb/slatedb/blob/main/src/db.rs#L421)

I don't think consistency for its own sake is a good enough reason. The WAL flush task freezes the active wal because the WAL flush is triggered by a time interval. On the other hand the memtable flush task is triggered by a size threshold. If we move freezing to the memtable flush task we'll get SSTs that are larger than the size threshold (depending on when the memtable flush task is able to react to the message).